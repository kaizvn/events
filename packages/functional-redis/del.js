const promisify = require('es6-promisify');

module.exports = promisify((client, key, cb) => client.del(key, cb));
