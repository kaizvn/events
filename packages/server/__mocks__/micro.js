const micro = jest.genMockFromModule('micro');

module.exports = micro;
module.exports.json = async (req) => {
  if (typeof req.body === 'object') {
    return req.body
  }

  throw new Error('Invalid JSON');
};
