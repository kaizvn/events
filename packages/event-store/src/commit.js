/* @flow */
import { createError } from 'micro-boom';
import { json } from 'micro';
import redis from 'redis';

import runLua from './runLua';

type Event = {
  type: string,
  payload: mixed,
};

const client = redis.createClient();

const commit = async (redisClient, event: Event) => {
  const lua = `
    local event = ARGV[1];
    local counter = redis.call('HLEN', KEYS[1]) + 1;

    redis.call('HSET', KEYS[1], counter, event);
    redis.call('PUBLISH', 'events', counter .. ':' .. event);
    return counter;
  `;

  return runLua(redisClient, lua, {
    keys: ['events'],
    argv: [JSON.stringify(event)],
  });
};

type TransformFunction = (any, http$IncomingMessage) => Event;

export default (transformReq: TransformFunction) => async (req: any) => {
  const body = await json(req);
  const event = transformReq(body, req);

  if (typeof event.type !== 'string') {
    throw createError(422, 'Missing type');
  }

  const id = await commit(client, event);

  return { ...event, id };
};
