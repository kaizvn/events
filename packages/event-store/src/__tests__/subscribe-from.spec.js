import del from 'redis-functional/del';
import micro from 'micro';
import range from 'lodash/fp/range';

import { delay } from 'awaiting';
import listen from 'test-listen';

import { commit } from '../commit';
import fromURL from './fromURL';
import makeSubscribe from '../subscribe';
import redisClient, { createClient } from '../redis-client';

jest.unmock('micro');

const takeEvents = (count, stream) =>
  stream.take(count).scan((prev, next) => prev.concat(next), []).toPromise();

describe('subscribe with last-event-id', () => {
  it('should send past event in cache', async () => {
    const pubClient = createClient();

    await del(redisClient, 'test-event-id::events');

    const { service, unsubscribe } = makeSubscribe({
      namespc: 'test-event-id',
      history: { size: 10 },
      burst: { time: 10, count: 10 },
    });
    const server = micro(service);
    const url = await listen(server);
    await delay(100);

    // commit 10 events (fitted in cache size)
    await Promise.all(
      range(1, 10).map(i =>
        commit(redisClient, { type: 'test', payload: i }, 'test-event-id')
      )
    );

    const headers = {
      'Last-Event-ID': '5',
    };

    const { data$, abort } = fromURL({ url, headers });

    // 1 -> :ok
    // 2 -> 6,7,8,9,10 events
    // 3 -> 11 real time
    // 4 -> abort
    const promise = takeEvents(4, data$);
    await delay(100);

    await commit(redisClient, { type: 'test', payload: 11 }, 'test-event-id');
    await delay(100);

    abort();
    server.close();
    unsubscribe();
    const val = await promise;
    pubClient.quit();

    expect(val).toMatchSnapshot();
  });
});
