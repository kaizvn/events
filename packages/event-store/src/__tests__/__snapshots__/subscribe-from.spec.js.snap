// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subscribe with last-event-id should send past event from redis if cache missed 1`] = `
Array [
  ":ok

",
  "id: 13
event: INCMSG
data: [{\\"id\\":4},{\\"id\\":5},{\\"id\\":6},{\\"id\\":7},{\\"id\\":8},{\\"id\\":9},{\\"id\\":10},{\\"id\\":11},{\\"id\\":12},{\\"id\\":13,\\"type\\":\\"test\\",\\"payload\\":1}]

",
  "id: 23
event: INCMSG
data: [{\\"id\\":14,\\"type\\":\\"test\\",\\"payload\\":2},{\\"id\\":15,\\"type\\":\\"test\\",\\"payload\\":3},{\\"id\\":16,\\"type\\":\\"test\\",\\"payload\\":4},{\\"id\\":17,\\"type\\":\\"test\\",\\"payload\\":5},{\\"id\\":18,\\"type\\":\\"test\\",\\"payload\\":6},{\\"id\\":19,\\"type\\":\\"test\\",\\"payload\\":7},{\\"id\\":20,\\"type\\":\\"test\\",\\"payload\\":8},{\\"id\\":21,\\"type\\":\\"test\\",\\"payload\\":9},{\\"id\\":22,\\"type\\":\\"test\\",\\"payload\\":10},{\\"id\\":23,\\"type\\":\\"test\\",\\"payload\\":11}]

",
  "id: 24
event: INCMSG
data: [{\\"id\\":24,\\"type\\":\\"test\\",\\"payload\\":12}]

",
]
`;

exports[`subscribe with last-event-id should send past event in cache 1`] = `
Array [
  ":ok

",
  "id: 15
event: INCMSG
data: [{\\"id\\":6},{\\"id\\":7},{\\"id\\":8},{\\"id\\":9},{\\"id\\":10},{\\"id\\":11,\\"type\\":\\"test\\",\\"payload\\":1},{\\"id\\":12,\\"type\\":\\"test\\",\\"payload\\":2},{\\"id\\":13,\\"type\\":\\"test\\",\\"payload\\":3},{\\"id\\":14,\\"type\\":\\"test\\",\\"payload\\":4},{\\"id\\":15,\\"type\\":\\"test\\",\\"payload\\":5}]

",
  "id: 19
event: INCMSG
data: [{\\"id\\":16,\\"type\\":\\"test\\",\\"payload\\":6},{\\"id\\":17,\\"type\\":\\"test\\",\\"payload\\":7},{\\"id\\":18,\\"type\\":\\"test\\",\\"payload\\":8},{\\"id\\":19,\\"type\\":\\"test\\",\\"payload\\":9}]

",
  "id: 20
event: INCMSG
data: [{\\"id\\":20,\\"type\\":\\"test\\",\\"payload\\":11}]

",
]
`;
